### PR TITLE
Adding channel POST endpoint.

### PR DESCRIPTION
**Pull request by MartinsIrbe**
+  Implemented channel translator from simple model to complex model.

### CODE REVIEW

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+import org.joda.time.LocalDate;
+
+public class ChannelModelTransformer implements ModelTransformer<Channel, org.atlasapi.media.channel.Channel> {
+
+    private final NumberToShortStringCodec v4Codec;
+    private final ImageModelTranslator imageTranslator;
+
+    public ChannelModelTransformer(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Code style. Constructor must be private. Make a static factory method named `create(...)` to create a new transformer

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    public org.atlasapi.media.channel.Channel transform(Channel simple) {
+        org.atlasapi.media.channel.Channel.Builder complex = org.atlasapi.media.channel.Channel.builder();
+
+        complex.withUri(simple.getUri());
+        complex.withHighDefinition(simple.getHighDefinition());
+        complex.withRegional(simple.getRegional());
+        complex.withAdult(simple.getAdult());
+        complex.withTitle(simple.getTitle());
+
+        Set<Alias> aliases = simple.getV4Aliases().stream().map(simpleAlias -> new Alias(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Break the method chain with newlines for readability

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                simpleAlias.getNamespace(),
+                simpleAlias.getValue()
+        )).collect(Collectors.toSet());
+        complex.withAliases(aliases);
+
+        if (simple.getMediaType() != null) {
+            Optional<MediaType> mediaType = MediaType.fromKey(simple.getMediaType().toUpperCase());
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Don't silently swallow the failure if `MediaType.fromKey()` returns absent. That means the simple model contains an invalid media type string. Throw a descriptive exception and fail the call.

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            complex.withAdvertiseFrom(new DateTime(simple.getAdvertisedFrom()));
+        }
+
+        if (simple.getGenres() != null && !simple.getGenres().isEmpty()) {
+            complex.withGenres(simple.getGenres());
+        }
+
+        if (simple.getPublisherDetails() != null) {
+            Maybe<Publisher> publisher = Publisher.fromKey(simple.getPublisherDetails().getKey());
+            if (!publisher.isNothing()) {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: We absolutely do not allow data to be written without a source. Same as above. If empty throw descriptive exception and fail the call

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                complex.withSource(publisher.requireValue());
+            }
+        }
+
+        if (simple.getBroadcaster() != null) {
+            Maybe<Publisher> publisher = Publisher.fromKey(simple.getBroadcaster().getKey());
+            if (!publisher.isNothing()) {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Ditto as above

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                complex.withBroadcaster(publisher.requireValue());
+            }
+        }
+
+        if (simple.getImages() != null && !simple.getImages().isEmpty()) {
+            complex.withImages(Iterables.transform(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Atlas Owl is now Java 8. `Iterables.transform(...)` should only be used if you need lazy evaluation which you don't here. Use Java 8 streams instead.

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                    simple.getImages(),
+                    input -> imageTranslator.transform(input)
+            ));
+        }
+
+        if (simple.getAvailableFrom() != null && !simple.getAvailableFrom().isEmpty()) {
+            complex.withAvailableFrom(Iterables.transform(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Ditto

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+        if (simple.getParent() != null) {
+            org.atlasapi.media.channel.Channel.Builder parent
+                    = org.atlasapi.media.channel.Channel.builder();
+            parent.withUri(simple.getParent().getUri());
+            complex.withParent(parent.build());
+        }
+
+        if (simple.getVariations() != null && !simple.getVariations().isEmpty()) {
+            complex.withVariations(Iterables.transform(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Ditto

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                    simple.getVariations(),
+                    channel -> transform(channel)
+            ));
+        }
+
+        simple.getRelatedLinks().stream().forEach(relatedLink -> {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Newlines to break up method chain

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            complex.withRelatedLink(translateRelatedLink(relatedLink));
+        });
+
+        if (simple.getChannelGroups() != null && !simple.getChannelGroups().isEmpty()) {
+            complex.withChannelNumbers(Iterables.transform(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Java 8 stream

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+        complex.withImage(simple.getImage());
+        complex.withShortName(simple.getShortName());
+        complex.withSourceId(simple.getSourceId());
+        complex.withThumbnail(simple.getThumbnail());
+
+        return complex.build();
+    }
+
+    private Publisher translatePublisherDetails(PublisherDetails simple) {
+        Maybe<Publisher> complex = Publisher.fromKey(simple.getKey());
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: Same as above. If you don't have a valid publisher throw exception

File: src/main/java/org/atlasapi/query/QueryWebModule.java [hunk 12]
```diff
@@ -365,6 +371,18 @@ private ContentWriteController contentWriteController() {
         );
     }
 
+    private ChannelWriteController channelWriteController() {
+        return new ChannelWriteController(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/QueryWebModule.java]: Code style. Use static factory methods or builders for all classes. Constructors should be private

File: src/main/java/org/atlasapi/query/v2/ChannelWriteController.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class ChannelWriteController {
+
+    private final Logger LOG = LoggerFactory.getLogger(ChannelWriteController.class);
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/ChannelWriteController.java]: Code style. Lowercase `log`

File: src/main/java/org/atlasapi/query/v2/ChannelWriteController.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    private final ApplicationConfigurationFetcher appConfigFetcher;
+    private final ChannelStore store;
+    private final ModelReader reader;
+    private final ChannelModelTransformer channelTransformer;
+
+    public ChannelWriteController(
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/ChannelWriteController.java]: Constructor should be private. Make a static factory method

File: src/main/java/org/atlasapi/query/v2/ChannelWriteController.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+import com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class ChannelWriteController {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/ChannelWriteController.java]: Missing `@Controller` annotation

File: src/main/java/org/atlasapi/query/QueryWebModule.java [hunk 16]
```diff
@@ -365,6 +371,18 @@ private ContentWriteController contentWriteController() {
         );
     }
 
+    private ChannelWriteController channelWriteController() {
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/QueryWebModule.java]: This should public with `@Bean` annotation

File: src/main/java/org/atlasapi/query/v2/QueryController.java [hunk 17]
```diff
@@ -210,4 +213,9 @@ public Void postContent(HttpServletRequest req, HttpServletResponse resp) {
     public Void putContent(HttpServletRequest req, HttpServletResponse resp) {
         return contentWriteController.putContent(req, resp);
     }
+
+    @RequestMapping(value="/3.0/channels.json", method = RequestMethod.POST)
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/QueryController.java]: Why is this here?
- First, this is the query controller for content. The POST endpoint for channels does not belong here
- Second, you have already defined this endpoint in its own controller where it should be.

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java (hunk 1 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: 👍

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java (hunk 2 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: 👍

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java (hunk 3 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: 👍

File: src/main/java/org/atlasapi/input/ChannelModelTransformer.java (hunk 4 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/input/ChannelModelTransformer.java]: 👍

File: src/main/java/org/atlasapi/query/v2/ChannelWriteController.java (hunk 15 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/query/v2/ChannelWriteController.java]: @Dalamar42 `ContentWriteController` doesn't have one either. Not sure if we really need this annotation.

File: src/main/java/org/atlasapi/query/QueryWebModule.java (hunk 16 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/query/QueryWebModule.java]: @Dalamar42 `ContentWriteController` is private and doesn't have @Bean annotation.

File: src/main/java/org/atlasapi/query/v2/QueryController.java (hunk 17 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/query/v2/QueryController.java]: @Dalamar42 I took an example from ContentWriteController. :) I have removed this as you are right it doesn't make any sense to have the annotation here as the entry point is in `QueryController` 👍

File: src/main/java/org/atlasapi/query/QueryWebModule.java (hunk 16 above)
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/QueryWebModule.java]: It looks like `ContentWriteController` is not declared properly and is intended defined inside the QueryController. This needs to be declared a bean and made public so it can be discovered by Spring

File: src/main/java/org/atlasapi/query/v2/ChannelWriteController.java (hunk 15 above)
Comment (Dalamar42) [inline src/main/java/org/atlasapi/query/v2/ChannelWriteController.java]: As below `ContentWriteController` is defined in a weird way. You absolutely need the `@Controller` annotation to ensure Spring discovers this class as a controller
